[query] return 0 if there was at least one valid number

gediminasgu opened the pull request:
**What this PR does / why we need it**:
PromQL compatibility fix.
When running instant queries PromQL returns "0" in place of NaN if there was at least one valid number in a range. Otherwise, it returns "NaN". That is the fix for M3 query to behave in the same way.

**Special notes for your reviewer**:
I'll change base branch to master when depending PR will be merged.

**Does this PR introduce a user-facing and/or backwards incompatible change?**:
```release-note
NONE
```

**Does this PR require updating code package or user-facing documentation?**:
```documentation-note
NONE
```

schallert committed:
[tracing] LightStep OpenTracing support (#2100)

- Add LightStep as a supported OT backend.
- Update docs.

robskillington committed:
[etcd] Upgrade etcd client library to 3.4.3 to fix data race in tests (#2101)

robskillington committed:
[dbnode] Fix TestIndexBlockFlush CI flakiness by allowing for longer stalls (#2102)

robskillington committed:
Move Makefile comment to avoid printing in CI logs (#2103)

robskillington committed:
[dbnode] Add filesystem index segment bootstrap benchmark test (#2104)

fingon committed:
[coordinator] Influxdb importer endpoint (at /api/v1/influxdb/write) (#2083)

notbdu committed:
Shareable Vagrant Scripts && Azure NVMe Disk Support (#2107)

* Add shared scripts and nvme disk provisioning code.

andrewmains12 committed:
Support custom grouping of instances by mirrored placement algorithm (#2074)

This PR makes the algorithm for grouping hosts into shardsets for the mirrored placement algorithm pluggable, and introduces a simple "group id" based strategy as an alternative The current behavior is:

Group instances by their port--e.g. 5555 with 5555, ensuring that:
a. Instances are on a different host and in a different isolation group
b. Instances have the same weight.

I've left the default behavior the same while introducing a new option, SetInstanceSelector to support changing the behavior.

arnikola committed:
[query] Fix a bug that could cause duplicate Prometheus label sets (#2112)

robskillington committed:
[dbnode] Add ability to format the host ID output with hostname resolver (#2109)

notbdu committed:
Use LVM && Modify Shared Scripts (#2111)

schallert committed:
[ci] Pull in latest coverage script (#2115)

schallert committed:
[kube] Update sysctl daemonset api path (#2117)

https://kubernetes.io/blog/2019/07/18/api-deprecations-in-1-16/

arnikola committed:
[query] Fix bug in graphite find result (#2116)

robskillington committed:
[dbnode] Allow returning empty term values from aggregate index values RPC (#2120)

schallert committed:
[ci] Pull in ci-scripts w/ package failures (#2121)

Commit:
Improved English (#2114)

gibbscullen committed:
(WIP Docs) Added tuning params to Prometheus remote write section  (#2118)

* added new remote write params

* added link to prom params

benraskin92 committed:
[docs] Add mapping rules config docs (#2124)

robskillington committed:
[dbnode] Add placement set handler for perform set operations (#2108)

justinjc committed:
Remove nested read lock to prevent deadlock (#2128)

* Remove nested read lock to prevent deadlock

* Make test into prop test and make it 'big'

* Respect shardIterateBatchMinSize

justinjc committed:
Make number of series a property of prop test (#2130)

* Make number of series a property of prop test

* More iterations and test larger batch size range

* Unbig shard_race_prop_test

notbdu committed:
Support Promremotebench Metrics && Add Monitoring Dash (#2131)

* Support promremotebench metrics.

gediminasgu committed:
[integrations] M3 Aggregator dashboard for Grafana (#2064)

schallert committed:
[kv] Include key name in watch errors (#2138)

arnikola committed:
Update operator version in developer dockerfile (#2137)

schallert committed:
[m3msg] Set key on value watches (#2144)

Allows for more verbose errors.

robskillington committed:
[coordinator] Delete related etcd keys when aggregator placement deleted (#2133)

notbdu committed:
[dbnode] Use Documents Builder And Persist Index Data In Peer BS (#2078)

* Use documents builder and share index persist logic in the peer bs path.

* [dbnode] Add mmap reporter (#2119)

* [dbnode] Skip shard insert queue for SeriesReadWriteRef (#2141)

* [dbnode] Improve speed writing postings offsets when writing FST segments (#2142)

Co-authored-by: Rob Skillington <rob.skillington@gmail.com>

fingon committed:
[coordinator] Influxdb write endpoint tag copy fix (#2126)

robskillington committed:
[changelog] Add changelog for 0.15.0-rc.0 release candidate (#2147)

notbdu committed:
Update kind & kube/prom operators, fix docker-integration-tests flakiness (#2149)

schallert committed:
[ci] Update ci-scripts (#2156)

benraskin92 committed:
[docs] Add upgrading m3 docs (#2154)

rallen090 committed:
[dbnode] Add trace coverage to FetchTagged code path (#2150)

schallert committed:
[ci] integration: "Install" netcat if needed (#2161)

We recently updated our Buildkite stack. The new version of the agents
doesn't come with netcat installed, which we use in our integration
tests. "Install" it if necessary.

schallert committed:
[query] api: Respect env+zone headers in topic (#2159)

justinjc committed:
[dbnode] Skip entries for unowned shards in commit log bootstrapper (#2145)

robskillington committed:
[dbnode] Fix bytes pool init logs (#2164)

justinjc committed:
[dbnode] Emit timer for bootstrap duration (#2163)

robskillington committed:
[dbnode] Add metrics for remote aggregator client and downsampler (#2165)

robskillington committed:
[dbnode] Add aggregator client maxBatchSize config for bigger buffer for data sent to aggregator (#2166)

robskillington committed:
[changelog] Add changelog for 0.15.0-rc.1 (#2167)

schallert committed:
[ci] Update ci-scripts for docker cleanup fix (#2168)

notbdu committed:
Align index block metric labels. (#2175)

notbdu committed:
Set defaults and expose configuration of tchannel timeouts. (#2173)

* Set defaults and expose configuration of tchannel timeouts.

notbdu committed:
Add regression test for and fix deadlock. (#2169)

schallert committed:
[kube] Add CAP_SYS_RESOURCE to m3db container (#2174)

arnikola committed:
[query] Add ability for custom series batching (#2181)

arnikola committed:
[dbnode] Add size approximation to series iterators (#2177)

robskillington committed:
[dbnode] [query] Add num series/bytes received and decode datapoint time to traces (#2158)

benraskin92 committed:
[docs] Add set placement endpoint docs (#2183)

simonrobb committed:
Async replication write success metric (#2180)

* Added replication success metric

* Updated .ci

* Typo

schallert committed:
[query] api: Allow in-place topic consumer updates (#2186)

Right now there are no HTTP APIs for updating a topic's consumers
in-place. The only way to do so would be deleting + recreating the
topic.

This adds support for `PUT`-ing a topic to update its consumers
in-place. Updating number of shards is currently not supported.

schallert committed:
[query] api: Fix agg+coord set+replace routes (#2189)

We had been serving on `/api/v1/${SVC}/set` rather than
`/api/v1/services/${SVC}/set`.

justinjc committed:
Don't return error on missing writable bucket during cold flush (#2188)

* Don't return error on missing writable bucket during cold flush

* Add clarifying comment

rallen090 committed:
[dbnode] Decode perf time ops (#2176)

Decode per time optimizations includes avoiding map access for time encoding schemes and using nanos int64 time operations instead of time struct

notbdu committed:
[dbnode] Concurrent time series indexing within a single batch (#2146)

* Concurrent indexing.

arnikola committed:
[dbnode] Flow segment checksum through TChannel (#2192)

arnikola committed:
[dbnode] Expose SchemaDescr on MultiReaderIterator (#2196)

arnikola committed:
[query] Fix a bug with step iteration allocating (#2185)

rallen090 committed:
[dbnode] Decode avoid checked bytes in segment reader (#2190)

notbdu committed:
[dbnode] Build a per field postings list (#2195)

* Create per field postings list.

robskillington committed:
[aggregator] Take last value by value timestamp not arrival time (#2199)

robskillington committed:
[dbnode] Add client bad request/internal error distinction for metrics and sampled logs (#2201)

robskillington committed:
[changelog] Update changelog for 0.15.0-rc.2 (#2200)

robskillington committed:
[goreleaser] Only build amd64 architecture for m3query releases (#2202)

jorgelbg committed:
[integrations] Add a datasource variable and reuse it in all the panels of the aggregator dashboard (#2182)

rallen090 committed:
[dbnode] Decode ReadBits improvements (#2197)

robskillington committed:
[coordinator] Ensure coordinator not grow M3Msg buffer if message over max size (#2207)

robskillington committed:
[development] Add HA Prometheus lab setup for dev M3 docker compose deployment (#2206)

robskillington committed:
[query] Fix namespace resolve debug log not being written with multiple namespaces (#2211)

justinjc committed:
Add metric for BootstrappedAndDurable (#2210)

robskillington committed:
[coordinator] Use tag options specified in config with M3Msg ingester (#2212)

notbdu committed:
[dbnode] Account for Neg/Pos Offsets when building per field roaring bitmap posting lists (#2213)

* Account for pos/neg offsets when building per field roaring bitmap posting lists.

tejasbhosale009 committed:
[coordinator] Add support for Graphite Grafana plugin /find POST requests (#2153)

notbdu committed:
[dbnode] Remove implicit cloning of time ranges to reduce Allocs (#2178)

* Remove implicit cloning of time ranges.

Co-authored-by: Rob Skillington <rob.skillington@gmail.com>

robskillington committed:
[coordinator] Only honor default aggregation policies if not matched by mapping rule (#2203)

* [coordinator] Only honor default aggregation policies if not matched by mapping rule

notbdu committed:
[changelog] Update changelog for 0.15.0-rc.3 (#2215)

rallen090 committed:
[dbnode] ReadBits to use optimized Read path (#2205)

andrewmains12 committed:
Add an in memory implementation of cluster/client.Client (#2219)

What this PR does / why we need it:

This adds a small testing utility to back a cluster/client.Client with an in-memory kv.Store. We have the cluster/kv/mem implementation of kv.Store, but there's currently no way to back a cluster/client.Client with it. Construction of the kv.Store is tightly coupled with etcd in the current class; having a separate class seemed like a reasonable way to do this given the current code, but we could consider a more dependency injection style for that--let me know if you want to hash that out.

schallert committed:
[dbnode] bootstrap: Make verbose log debug lvl (#2222)

gibbscullen committed:
[docs] How to write to multiple M3DB clusters (#2187)

* added new remote write params

* added link to prom params

* adding section for multiple m3db clusters

* first round of edits made

* small edits and additions

* formatting changes

* final changes

* nit

robskillington committed:
[query] Log query errors with query params (#2225)

notbdu committed:
Add Support For Index Volume Types (#2223)

* Add index volume types and default index volume type.

robskillington committed:
[query] Fix query limit metrics reporting not starting at startup (#2228)

robskillington committed:
[dbnode] Use invariant error for flush errors to fail CI with flush errors in integration tests (#2217)

* [dbnode] Fail builds with flush errors in integration tests

* [dbnode] DB Close Wait for Fs Processes (#2229)

* [dbnode] Only read data from disk when flushing index segments.

arnikola committed:
[query] add an endpoint to ingest results (#2224)

cw9 committed:
Make errBufferFull public (#2237)

schallert committed:
[query] Noop storage backend to serve admin APIs (#2236)

- Some use cases require the coordinator for serving m3admin APIs but
  have no need to communicate with M3DB or serve query data.
- Currently need a live M3DB cluster in order to serve m3admin APIs.
- This PR adds a `noop-etcd` storage backend for serving m3admin APIs.

Example of querying:
```
$ curl -v "localhost:7201/api/v1/query_range?start=$(date '+%s')&end=$(date '+%s')&step=10&query=foo"
*   Trying ::1...
* TCP_NODELAY set
* Connected to localhost (::1) port 7201 (#0)
> GET /api/v1/query_range?start=1585603982&end=1585603982&step=10&query=foo HTTP/1.1
> Host: localhost:7201
> User-Agent: curl/7.54.0
> Accept: */*
>
< HTTP/1.1 500 Internal Server Error
< Access-Control-Allow-Headers: accept, content-type, authorization
< Access-Control-Allow-Methods: POST, GET, OPTIONS, PUT, DELETE
< Access-Control-Allow-Origin: *
< Date: Mon, 30 Mar 2020 21:33:02 GMT
< Content-Length: 48
< Content-Type: text/plain; charset=utf-8
<
{"error":"operation not valid for noop client"}
```

schallert committed:
[ci] Temporarily disable kubeval validation (#2241)

Kubeval depends on `hashicorp/go-multierror`, which now depends on
go1.13 error methods.

arnikola committed:
[query] Fix temporal function tag application bug (#2231)

robskillington committed:
[dbnode] Use madvdontneed=1 for GODEBUG flags (#2242)

arnikola committed:
[changelog] Update changelog for 0.15.0-rc.4 (#2243)

richardpoole committed:
[docs] Add StorageClass volumeBindingMode guidance (#2234)

Co-authored-by: Matt Schallert <mattschallert@gmail.com>

notbdu committed:
[dbnode] ShardTimeRanges AddRanges Clone if Not Exists (#2246)

andrewmains12 committed:
Use random ports for etcd (#2244)

What this PR does / why we need it:

Currently, our embedded KV module uses etcd's default ports. This is problematic in a CI environment; we can't guarantee openness of those ports. Better is to use a random port, which gives us test isolation.

Note: if there's some code assuming the default ports, this will break it. I'm not concerned about that because:

This is a change to a test module.
Users can already reference the endpoints using Endpoints()
If we don't want the backwards incompatible change though, I can put it behind a flag.
Special notes for your reviewer:

Does this PR introduce a user-facing and/or backwards incompatible change?:

Changes src/cluster/integration to use random ports for etcd. The server addresses can be accessed via `Endpoints()`.
Does this PR require updating code package or user-facing documentation?:

NONE

arnikola committed:
[query] Fix histogram grouping bug (#2247)

schallert committed:
[opentracing] Fix typo in config error (#2248)

robskillington committed:
[coordinator] Add headers to along with request to remote write forward target (#2249)

robskillington committed:
[changelog] Squash changelog for 0.15 unreleased and update for rc.5 (#2250)

arnikola committed:
[dbnode] Add instrumentation to forwards index (#2252)

schallert committed:
[coordinator] Forward all relevant M3 headers to forwarding targets (#2256)

robskillington committed:
[query] Add ability to set default query timeout by config (#2226)

arnikola committed:
[query] Fix offset fetching too much data (#2260)

robskillington committed:
[aggregator] Add compatibility for rollup rules with timed metrics (#2251)

schallert committed:
[coordinator] Support mapping write tags via a header (#2255)

Co-authored-by: Rob Skillington <rob@chronosphere.io>

notbdu committed:
Expose IndexVolumeType of IndexFileSetReader. (#2258)

* Expose IndexVolumeType of IndexFileSetReader.

arnikola committed:
[query] Allow generalization for rate queries. (#2263)

Commit:
[aggregator] Add passthrough functionality in m3aggregator using rawtcp server (#2235)

* Add passthrough functionality in m3aggregator using rawtcp server

The m3aggregator instances will take passthrough metrics (with timestamp and storage policy), and directly write them to the ingestion tier.

notbdu committed:
[dbnode] Add on cold flush processor (#2266)

* [dbnode] Add on cold flush processor

Co-authored-by: Rob Skillington <rob@chronosphere.io>

xmcqueen committed:
[tools] CLI tool to simplify some tasks, and eliminate the curl commands from the docs (#2097)

robskillington committed:
Add first draft of roadmap (#2122)

schallert committed:
[aggregator] Allow static shutdown wait period (#2264)

notbdu committed:
[dbnode] Cleanup duplicate index segments from disk (#2261)

* [dbnode] Add code to cleanup dupe index segments from disk.

wybczu committed:
Fix m3ctl README formatting (#2269)

rallen090 committed:
[dbnode] Track blocks recently queried within some duration in dbnode (#2240)

BertHartm committed:
[grafana] Fix refId for Persistence graph in M3DBNode dashboard (#2271)

Having 2 refIds that are identical in the same panel causes grafana to be unable to load the panel. This sets the 2nd query to be a different refId

arnikola committed:
[query] Refactor query code, add warnings to prom output (#2265)

vdarulis committed:
Wire up BatchFlushDeadline in aggregator client config (#2273)

notbdu committed:
[dbnode] Decouple index builder w/ lifetime of fssource, alloc every read() call (#2274)

robskillington committed:
[aggregator] Add M3Msg client and server for M3Aggregator (#2171)

notbdu committed:
[dbnode] Fix dedupe/superset logic (#2275)

* Fix dedupe/superset logic.

rishabhjain08 committed:
[dbnode] Prevent repairs from getting stuck on a bad block (#2272)

* Prevent repairs from getting stuck on a bad block

- Each namespace is repaired in decreasing time order i.e. from future
to past. Currently, if a namespace repair fails for some time range T,
then repair for all time ranges <T is skipped until the next time
database repair is scheduled on the node.

- This commit changes the above mentioned behavior by allowing database
repairer to find and repair time range just before T which needs repair,
if repairiing time range T failed.

- Note that if no time range is repaired either because time range
didnt need to be repaired or because repairing a time range failed,
then LRR (least recently repaired) policy is used to repair LRR time
range. In some scenarios this can cause database repairer to repair
same time range twice; once during time ordered repair for namespace
and second time while attempting to repair LRR time range.

- This commit introduces a unit test to test all the above mentioned
behavior changes.

* Addresses comments on PR and updates unit tests

arnikola committed:
[query] Remote debug endpoint for remote read (#2276)

prateek committed:
[cluster] Add InstanceMetadata per placement.Instance (#2279)

notbdu committed:
[dbnode] Add cold writes override (#2280)

* [dbnode] allow force cold writes

arnikola committed:
[query] Add parse expression functionality (#2278)

notbdu committed:
[dbnode] Allow configuring of a write timestamp offset into the past. (#2281)

notbdu committed:
[dbnode] Persist manager close resources (#2282)

robskillington committed:
[changelog] Update changelog for unreleased 0.15 ahead of RC.6 (#2283)

robskillington committed:
[m3msg] Use multiple connections for M3Msg writers (#2230)

notbdu committed:
Make sure writeoffset is always set. (#2284)

notbdu committed:
Decouple index builder from lifetime of peers bootstrapper. (#2285)

robskillington committed:
[dbnode] Add tracing for bootstrap process (#2216)

robskillington committed:
[query] Add instrument options ready channel to receive instrument opts (#2286)

robskillington committed:
[coordinator] Allow configuring of matcher configuration (#2289)

notbdu committed:
[dbnode] Change Bootstrap Hooks to Hook Iface (#2296)

rallen090 committed:
Add retry to write forwarding (#2299)

robskillington committed:
[dbnode] Use histogram by default for instrument.NewTimer in DB node (#2302)

robskillington committed:
[dbnode] Remove pooling config since defaults now included (#2297)

This was drifting from the defaults and people had been copying it unfortunately, so removing this from the config here.

arnikola committed:
[dbnode] Made result pools optional, allocing otherwise (#2301)

robskillington committed:
[dbnode] Use instrument default histogram buckets not tally (#2303)

arnikola committed:
[dbnode] No allocs for term only aggregate queries (#2304)

arnikola committed:
[logging] Update types on some log fields (#2306)

Commit:
[m3comparator] a feature to clear data on memory storage

Commit:
test added

arnikola committed:
[query] Add regression tests to comparator CI (#2290)

Commit:
Merge branch 'master' into gg/m3comparator-clear-data

arnikola committed:
[query] Debug endpoint no longer dedupes tags (#2314)

robskillington committed:
[coordinator] Return 4XX on duplicate tags or tags w/ no name for remote write (#2315)

robskillington committed:
[dbnode] Add verbose out of retention errors (#2317)

linasm committed:
[comparator] Generate random multi series metrics

linasm committed:
Remove unnecessary comment

linasm committed:
Address review feedback

arnikola committed:
Merge branch 'master' into gg/m3comparator-clear-data

linasm committed:
[comparator] Filter query results by tag matchers

Commit:
Merge remote-tracking branch 'origin/linasm/comparator-filter-by-tags' into gg/m3comparator-clear-data

gibbscullen committed:
Removing link to Gitter (#2320)

We are shutting down the Gitter channel so removed the link from the Read me.

robskillington committed:
[dbnode] Add ability to write existing FST segment out (#2321)

notbdu committed:
[dbnode] Ignore Dupe ID Errors Earlier and Add Metrics/Error Logs to Shard Insert Queue (#2322)

Co-authored-by: Rob Skillington <rob@chronosphere.io>

notbdu committed:
Don't perform any cleanup while we are still bootstrapping. (#2324)

notbdu committed:
Properly handle cold index errors if cold writes are enabled. (#2325)

rishabhjain08 committed:
[dbnode] Allow reading/writing log entries greater than flushMaxBytes from/to commit log (#2277)

* Allows writing log entries bigger than flush size in commit log

- Resolves a bug in chunk writer which resulted bufio to skip wrong
number of bytes while trying to perform a buffered write of commit log
entry.

- Resolves an issue an commit log reader allowing commit log reader to
read chunk's greater than flush size. Chunk reader was trying to peek
data in the chunk to compute its checksum. This resulted in buffer full
error for large chunk sizes. This commit resolves this issue by loading
the chunk data into a new buffer to compute its checksum instead of
trying to peek  for data in bufio's fixed length buffer.

- Introduces unit tests to test for all the above mentioned fixes with
different filled buffer capacities to make sure chunk reader/write can
gather log entries of any size across buffers and chunks without any
error.

wybczu committed:
[docs] Cleanup README files (#2326)

justinjc committed:
Fix upserting consecutive protobuf writes (#2323)

* Don't skip encoding if both value and annotation is the same as previous

arnikola committed:
[query] Allow custom storage transform (#2327)

arnikola committed:
[query] Update backend storage transform interface (#2330)

robskillington committed:
[dbnode] Sort by unmarked index entries for indexing (#2328)

Commit:
[m3comparator] Querying without metric name

Allows to run queries like {tag="test"}. Previously only querying with metric name like random{tag="test"} was only supported.

schallert committed:
Merge branch 'master' into m3comparator-query-wo-metric-name

Commit:
promql testdata migrated to m3 stack

linasm committed:
[comparator] Generate random multi series metrics (#2318)

notbdu committed:
[dbnode] Track when a series bucket was first written. (#2333)

notbdu committed:
[dbnode] Skip index queue for cold writes (#2329)

notbdu committed:
[dbnode] Ensure block leasers are eventually consistent w/ shard state (#2338)

* Rip out block retriever from bootstrappers.

robskillington committed:
[dbnode] Adjust default pool sizes based on workloads at various organizations (#2337)

notbdu committed:
[dbnode] Fix first write logic (#2341)

Commit:
test files added

arnikola committed:
[query] Fix metrics for  forwarding latency (#2342)

schallert committed:
[ci] Update ci-scripts w/ codecov changes (#2350)

notbdu committed:
[dbnode] Refactor index block into mutable segments and block (#2345)

notbdu committed:
[dbnode] Add cold mutable segs and expose APIs to rotate/evict (#2348)

notbdu committed:
[dbnode] Operate upon on cold mutable segments in index block (#2352)

gibbscullen committed:
Add meetup group for community meetings to README.md (#2353)

* Update README.md

* Update README.md

* Update README.md

Co-authored-by: Rob Skillington <rob.skillington@gmail.com>

notbdu committed:
[dbnode] Let cold writes through and seal cold created index blocks (#2354)

rishabhjain08 committed:
[dbnode] Emits metrics to monitor divergence of shards as db repair progresses (#2334)

robskillington committed:
[coordinator] Add multi-process config for on-prem process scaling (#2292)

linasm committed:
[query] Set warning headers in remote read debug endpoint (json format) (#2356)

linasm committed:
[comparator] Filter query results by tag matchers (#2319)

Commit:
[query] Native PromQL integration into m3coordinator (#2343)

Commit:
[m3comparator] a feature to clear data on memory storage (#2312)

Commit:
Merge remote-tracking branch 'origin/master' into m3comparator-query-wo-metric-name

# Conflicts:
#	src/cmd/services/m3comparator/main/filterer.go
#	src/cmd/services/m3comparator/main/main.go
#	src/cmd/services/m3comparator/main/querier.go
#	src/cmd/services/m3comparator/main/querier_test.go

Commit:
Merge branch 'master' into m3comparator-query-wo-metric-name

# Conflicts:
#	src/cmd/services/m3comparator/main/series_load_handler.go
#	src/cmd/services/m3comparator/main/series_load_handler_test.go

Commit:
querier test refactored

shreyassrivatsan committed:
Add metric type and source to timeseries message (#2359)

Co-authored-by: Jerome Froelich <jeromefroelich@hotmail.com>

linasm committed:
[query] Fix tag mismatch message text (#2360)

Co-authored-by: Rob Skillington <rob.skillington@gmail.com>

linasm committed:
[dbnode] Allow construction of QueryStatsTracker to use instrument opts (#2358)

arnikola committed:
[thrift] Add metadata to node health checks (#2362)

robskillington committed:
[coordinator] Add write/read retry calls vs final-error metrics by default (#2365)

arnikola committed:
[dbnode] Clean up new client connection generator (#2363)

arnikola committed:
[thrift] Hotfix and test for service metadata. (#2366)

robskillington committed:
[dbnode] Add ability to configure peer streaming shard concurrency (#2364)

arnikola committed:
[dbnode] Add option for experimental TChannel server use (#2367)

robskillington committed:
[query] Add Prometheus engine metrics and upgrade to Prometheus 2.17.2 (#2369)

Commit:
Merge remote-tracking branch 'origin/master' into gg/m3comparator-promql-testdata

# Conflicts:
#	src/cmd/services/m3comparator/main/filterer.go
#	src/cmd/services/m3comparator/main/main.go
#	src/cmd/services/m3comparator/main/querier.go
#	src/cmd/services/m3comparator/main/querier_test.go
#	src/cmd/services/m3comparator/main/series_load_handler.go
#	src/cmd/services/m3comparator/main/series_load_handler_test.go

Commit:
fixes

shreyassrivatsan committed:
[coordinator] Support additional metric types on remote write (#2368)

linasm committed:
[query] Support parsing/matching of non matrix results (#2374)

Commit:
Merge remote-tracking branch 'origin/master' into gg/m3comparator-promql-testdata

Commit:
Merge branch 'master' into m3comparator-query-wo-metric-name

Commit:
fixes

linasm committed:
[dbnode] Fix query_stats_total_docs_per_block counter (#2357)

Commit:
trigger

Commit:
Merge branch 'm3comparator-query-wo-metric-name' into gg/m3comparator-promql-testdata

# Conflicts:
#	src/cmd/services/m3comparator/main/querier.go
#	src/cmd/services/m3comparator/main/querier_test.go

Commit:
[m3comparator] query without metric name (#2335)

Commit:
Merge remote-tracking branch 'origin/master' into gg/m3comparator-promql-testdata

# Conflicts:
#	src/cmd/services/m3comparator/main/querier.go
#	src/cmd/services/m3comparator/main/querier_test.go

Commit:
fixes

linasm committed:
[comparator] Support generating random histogram metrics (#2379)

Commit:
Merge remote-tracking branch 'origin/master' into gg/m3comparator-promql-testdata

# Conflicts:
#	src/cmd/services/m3comparator/main/querier_test.go

notbdu committed:
Remove Evict() API and call Close(). (#2382)

cw9 committed:
[msg] Fix metrics (#2372)

Commit:
test run fix

notbdu committed:
[all] Upgrade Go runtime to 1.13, use go modules, remove glide and use MADVISE=dontneed for dbnode (#2221)

gediminasgu committed:
Update src/query/test/testdata/histograms.test

Co-authored-by: Linas Medžiūnas <linasm@users.noreply.github.com>

gediminasgu committed:
Update src/query/test/testdata/histograms.test

Co-authored-by: Linas Medžiūnas <linasm@users.noreply.github.com>

gediminasgu committed:
Update src/query/test/testdata/literals.test

Co-authored-by: Linas Medžiūnas <linasm@users.noreply.github.com>

Commit:
minor changes

Commit:
Merge remote-tracking branch 'origin/master' into gg/m3comparator-promql-testdata

Commit:
Merge branch 'gg/m3comparator-promql-testdata' of github.com:gediminasgu/m3 into gg/m3comparator-promql-testdata

Commit:
[query] Return M3-Results-Limited header when series limit is reached for Prometheus query engine (#2371)

robskillington committed:
[changelog] Update changelog for 0.15 release (#2387)

linasm committed:
[query] Sort grouped aggregate outputs by tags (#2386)

arnikola committed:
[dtest] Initial docker integration test migration (#2340)

notbdu committed:
[dbnode] Add documentation on flushing/snapshotting (#2385)

dependabot[bot] committed:
Bump websocket-extensions from 0.1.3 to 0.1.4 in /src/ctl/ui (#2394)

Commit:
Merge remote-tracking branch 'origin/master' into gg/m3comparator-promql-testdata

Commit:
minor changes

Commit:
[query] return 0 if there was at least one valid number

robskillington committed:
[build] Build release binaries with goreleaser using Go 1.13 (#2397)

robskillington committed:
[coordinator] Ingesting aggregated metrics with invalid params should always log error (#2398)

linasm committed:
[query] Exclude metrics name from without-based aggregations (#2396)

linasm commented:
I think this will clash with https://github.com/m3db/m3/pull/2390. Maybe there is a way to solve this at a deeper layer than the rendering of the output?

schallert committed:
[m3admin] API: Support namespace retention updates (#2383)

We currently don't have a way to modify a namespaces retention with the
m3admin APIs. However doing so is somewhat tricky, as it's unsafe to
update any arbitrary namespace field.

This PR supports safely updating a namespace in-place. The only
respected field is retention period, as we know that to be safe.
Attempts to modify any other field will inform the user the field is
immutable.

linasm committed:
[query] Respect keepNans option for instant queries (#2390)

linasm committed:
[dbnode] Report a histogram of loaded docs per query (#2381)

linasm committed:
[query] Workaround for non monotonic histogram buckets (#2388)

Commit:
[m3comparator] adaption of Prometheus testdata to validate m3query (#2361)

Commit:
Merge remote-tracking branch 'origin/master' into gg/only-nan-fix

# Conflicts:
#	src/query/api/v1/handler/prometheus/native/common.go
#	src/query/test/m3comparator_client.go
#	src/query/test/m3query_client.go
#	src/query/test/test.go
#	src/query/test/testdata/functions.test
#	src/query/test/testdata/histograms.test
#	src/query/test/testdata/legacy.test

Commit:
rollback of some changes

Commit:
trigger
